<a href="https://colab.research.google.com/github/Saldebusts/sql_database/blob/main/SQL_Data_base.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Capital Bikeshare y el tiempo en Washington D. C. (2011–2012)

## Contexto y objetivo

Nos ponemos en el lugar de una empresa de bicicletas compartidas que necesita organizar su servicio según los momentos de mayor y menor demanda.

Nuestra pregunta principal es: **¿cuándo conviene reforzar el servicio de bicicletas y cuándo programar el mantenimiento?**

Analizamos los alquileres de Capital Bikeshare durante 2011 y 2012 y su relación con el calendario y las condiciones meteorológicas. Buscamos identificar patrones que nos ayuden a proponer horarios de refuerzo y periodos de menor demanda candidatos para mantenimiento.

## Fuentes de datos

| Fuente | Formato | Qué aporta |
|---|---|---|
| UCI Bike Sharing Dataset | CSV: `day.csv` y `hour.csv` | Alquileres diarios y horarios, separados entre ocasionales y registrados, junto con información del calendario y meteorología. |
| Open-Meteo — Historical Weather API | API: JSON | Estimaciones históricas de temperatura media diaria, precipitación acumulada y viento máximo diario. |

Las dos fuentes corresponden a Washington D. C. y al periodo 2011–2012. Las relacionamos mediante la fecha.

## Preguntas del análisis

1. ¿Qué horas y tipos de día concentran más alquileres?
2. ¿Qué diferencias encontramos entre los patrones de alquiler de usuarios ocasionales y registrados?
3. ¿Cómo cambia la media diaria de alquileres según la precipitación? ¿La diferencia es igual para ambos tipos de usuario?
4. ¿Cómo varían los alquileres según la temperatura?
5. ¿Cómo evoluciona la demanda entre 2011 y 2012?

## Hipótesis

- **H1 — Horarios y tipo de usuario:** en días laborables, los alquileres de usuarios registrados presentan picos alrededor de las 8:00 y de las 17:00–18:00. Los ocasionales presentan una mayor media de alquileres por hora durante las tardes y los fines de semana.

- **H2 — Precipitación:** los días con al menos 1 mm de precipitación presentan una menor media diaria de alquileres que los días por debajo de ese umbral. La diferencia porcentual es mayor entre los ocasionales que entre los registrados.

- **H3 — Evolución anual:** la media diaria de alquileres aumenta entre 2011 y 2012, con un crecimiento porcentual mayor entre los registrados que entre los ocasionales.

- **H4 — Temperatura:** la media diaria de alquileres aumenta con la temperatura hasta cierto nivel y disminuye en los días de temperaturas más altas.

Estas hipótesis son propuestas que contrastaremos con los datos, no conclusiones previas. Para comparar las condiciones meteorológicas, tendremos en cuenta el año y el tipo de día.

## Indicadores principales

- Total y media diaria de alquileres.
- Media de alquileres por hora y tipo de día.
- Porcentaje de alquileres de usuarios registrados.
- Diferencia porcentual de alquileres según la precipitación.
- Media diaria de alquileres por rangos de temperatura.
- Variación porcentual de la demanda entre años.

## Alcance del proyecto

Estudiamos el conjunto del sistema de bicicletas compartidas. Los recuentos representan alquileres, no personas diferentes.

Los datos actuales permiten analizar cuándo se utiliza el servicio, pero no identificar estaciones vacías o llenas, rutas o duración de los viajes. Dejamos el análisis de reubicación entre estaciones como una posible ampliación.

Usamos la meteorología de Open-Meteo para las comparaciones diarias. Las condiciones meteorológicas por hora proceden de UCI.

## Qué hacemos en este notebook

Extraemos las dos fuentes, revisamos su calidad y comprobamos que se pueden relacionar por la fecha. Después preparamos cuatro tablas, definimos sus claves y las cargamos en MySQL.

Las tablas son:

- `calendario`: características de cada día.
- `meteorologia_dia`: condiciones meteorológicas diarias de Open-Meteo.
- `situacion_meteo`: significado de los códigos meteorológicos de UCI.
- `alquileres_hora`: alquileres y condiciones meteorológicas de cada registro horario.

Esta base de datos servirá para responder las preguntas del proyecto mediante consultas SQL y gráficos en el notebook de análisis.

## Organización del notebook

- **Parte 1 · Extracción y calidad de los datos:** secciones 1 a 13.
- **Parte 2 · Limpieza y base de datos MySQL:** secciones 14 a 23.

## Cómo ejecutarlo

Ejecutamos las celdas en orden en Google Colab.

La sección 19 instala y arranca MySQL en el entorno de Colab. La sección 20 recrea las tablas, por lo que borra su contenido anterior. Después, la sección 21 carga los datos preparados.

La sección 23 solicita acceso a Google Drive para guardar los archivos limpios, el esquema SQL y una exportación de la base de datos.

## Parte 1 · Extracción y calidad de los datos

### 1. Preparamos las herramientas y el pedido a la API

**Qué hacemos:** importamos `requests`, que hace peticiones a internet, y `pandas`, que organiza los datos en tablas. Después escribimos la dirección de la API de Open-Meteo y los parámetros del pedido.

**Por qué la dirección y los parámetros por separado:** una URL con todo escrito seguido es difícil de leer y fácil de romper: basta con olvidar un `&`. En un diccionario se ve qué pedimos y se cambia un valor sin tocar el resto; `requests` monta la URL completa.

**Qué pedimos:**
- `latitude` y `longitude`: el centro de Washington D. C., donde funciona Capital Bikeshare.
- `start_date` y `end_date`: 2011 y 2012, el mismo periodo que los datos de bicicletas.
- `daily`: tres variables diarias: temperatura media, precipitación total y viento máximo.
- `timezone`: obligatorio para datos diarios. Con `America/New_York` cada día va de 00:00 a 24:00 en hora de Washington, como las fechas de UCI. Sin él, los días empezarían a medianoche UTC, que en Washington son las 19:00 o las 20:00 del día anterior.

In [ ]:
import requests
import pandas as pd

In [ ]:
# Dirección de la API y parámetros del pedido, por separado.
url_api = "https://archive-api.open-meteo.com/v1/archive"

parametros = {
    "latitude": 38.8951,             # Washington D. C.
    "longitude": -77.0364,
    "start_date": "2011-01-01",      # mismo periodo que los datos de bicicletas
    "end_date": "2012-12-31",
    "daily": "temperature_2m_mean,precipitation_sum,wind_speed_10m_max",
    "timezone": "America/New_York",  # cada día va de 00:00 a 24:00 en hora de Washington
}
parametros

{'latitude': 38.8951,
 'longitude': -77.0364,
 'start_date': '2011-01-01',
 'end_date': '2012-12-31',
 'daily': 'temperature_2m_mean,precipitation_sum,wind_speed_10m_max',
 'timezone': 'America/New_York'}

### 2. Solicitamos los datos meteorológicos

**Qué hacemos:** enviamos la petición con `requests.get`, pasando la dirección y los parámetros, y comprobamos la respuesta.

**Por qué el `timeout`:** si el servidor no responde, sin `timeout` la celda se quedaría esperando para siempre. Con 60 segundos, falla con un error claro.

**Por qué comprobar el código de estado:** 200 significa que todo ha ido bien; 400, que algún parámetro está mal; 500, que el error es del servidor. Si no es 200, Open-Meteo explica el motivo en la clave `reason`, y `raise_for_status()` detiene el notebook para no seguir con datos vacíos.

**Por qué no imprimimos `response.content`:** es la respuesta entera en bytes, decenas de miles de caracteres. No se puede leer y ocupa la pantalla; en la sección 3 la convertimos en un diccionario y miramos solo lo que necesitamos.

In [ ]:
response = requests.get(url_api, params=parametros, timeout=60)
print(response.status_code)
print(response.url)

200
https://archive-api.open-meteo.com/v1/archive?latitude=38.8951&longitude=-77.0364&start_date=2011-01-01&end_date=2012-12-31&daily=temperature_2m_mean%2Cprecipitation_sum%2Cwind_speed_10m_max&timezone=America%2FNew_York


In [ ]:
# Si algo falla, Open-Meteo explica el motivo en la clave "reason".
if response.status_code != 200:
    print("Motivo:", response.json().get("reason"))
response.raise_for_status()   # detiene el notebook si la petición ha fallado
print("Petición correcta")

Petición correcta


**Qué leemos:** código 200 y "Petición correcta". La URL que muestra `response.url` es la que ha montado `requests`: se puede copiar y abrir en el navegador para ver el mismo JSON. Las comas aparecen como `%2C` y la barra de la zona horaria como `%2F`: es la forma en que las URL codifican esos caracteres.

### 3. Leemos la respuesta de la API

**Qué hacemos:** convertimos la respuesta JSON en un diccionario de Python con `.json()`, miramos sus claves y comprobamos las unidades en `daily_units`.

**Por qué `.json()` y no `response.text`:** `text` es una cadena de texto corrida; `.json()` la convierte en un diccionario en el que se llega a cada valor por su nombre, por ejemplo `data["daily"]`.

**Por qué mirar las unidades:** no hay que suponerlas. Otras APIs dan la temperatura en Fahrenheit o el viento en m/s; si nos equivocamos aquí, todas las conclusiones sobre el tiempo estarían mal.

In [ ]:
data = response.json()
print(data.keys())

dict_keys(['latitude', 'longitude', 'generationtime_ms', 'utc_offset_seconds', 'timezone', 'timezone_abbreviation', 'elevation', 'daily_units', 'daily'])


In [ ]:
data["daily_units"]

{'time': 'iso8601',
 'temperature_2m_mean': '°C',
 'precipitation_sum': 'mm',
 'wind_speed_10m_max': 'km/h'}

**Qué leemos:** el diccionario tiene nueve claves. Las que nos interesan son `daily_units` (las unidades) y `daily` (los datos); las demás describen el pedido: coordenadas, zona horaria, altitud y el tiempo que tardó el servidor.

Unidades: temperatura en °C, precipitación en mm y viento en km/h. Las fechas vienen en formato ISO 8601 (`AAAA-MM-DD`).

**Aviso:** Open-Meteo devuelve latitud 38,91 y longitud -77,07, no exactamente las que pedimos. Es el punto de su cuadrícula más cercano a Washington: los datos salen de un modelo meteorológico con celdas de varios kilómetros, no de una estación concreta.

### 4. Convertimos los datos diarios en una tabla

**Qué hacemos:** seleccionamos `daily` dentro del diccionario y lo convertimos en un DataFrame. Cada fila será un día, con su fecha, temperatura media, precipitación y viento máximo.

**Por qué funciona en una línea:** `daily` es un diccionario de listas, una lista por variable y todas del mismo tamaño (una posición por día). Es exactamente el formato que espera `pd.DataFrame`: cada lista se convierte en una columna.

**Por qué no un bucle `for`:** recorrer el JSON día a día para montar las filas daría el mismo resultado con diez líneas más y más ocasiones de equivocarse.

In [ ]:
weather_df = pd.DataFrame(data["daily"])
weather_df.head()

,time,temperature_2m_mean,precipitation_sum,wind_speed_10m_max
0,2011-01-01,8.2,2.5,17.7
1,2011-01-02,8.6,3.0,18.9
2,2011-01-03,-0.1,0.0,17.4
3,2011-01-04,0.1,0.0,11.2
4,2011-01-05,-0.6,0.0,13.9


**Qué leemos:** cuatro columnas: `time` (la fecha, todavía como texto), `temperature_2m_mean`, `precipitation_sum` y `wind_speed_10m_max`. El 1 y el 2 de enero de 2011 fueron suaves (unos 8 °C) y con algo de lluvia (2,5 y 3 mm); desde el día 3 la temperatura baja a cero.

### 5. Comprobamos el periodo, los valores ausentes y los extremos

**Qué hacemos:** revisamos el tamaño de la tabla, la primera y la última fecha, los valores nulos de cada columna y, con `describe()`, el mínimo, el máximo y la media de cada variable. Localizamos también el día más lluvioso.

**Por qué:** esperamos 731 filas, 365 días de 2011 y 366 de 2012, que fue bisiesto. Si falta un día, habrá un hueco al unir con los datos de bicicletas. `describe()` sirve para detectar valores imposibles, como una temperatura de 60 °C o una precipitación negativa.

**Por qué `idxmax()`:** devuelve la posición de la fila con el valor más alto; con `loc` recuperamos esa fila entera, con su fecha.

In [ ]:
print("Filas y columnas:", weather_df.shape)

print("Primera fecha:", weather_df["time"].min())
print("Última fecha:", weather_df["time"].max())

print("Valores nulos:")
print(weather_df.isna().sum())

Filas y columnas: (731, 4)
Primera fecha: 2011-01-01
Última fecha: 2012-12-31
Valores nulos:
time                   0
temperature_2m_mean    0
precipitation_sum      0
wind_speed_10m_max     0
dtype: int64


In [ ]:
display(weather_df.describe().round(1))

print("Día más lluvioso:")
print(weather_df.loc[weather_df["precipitation_sum"].idxmax()])

,temperature_2m_mean,precipitation_sum,wind_speed_10m_max
count,731.0,731.0,731.0
mean,14.7,2.9,15.3
std,9.3,7.4,5.7
min,-5.2,0.0,6.4
25%,6.5,0.0,11.2
50%,15.0,0.0,13.9
75%,22.9,2.1,18.3
max,34.0,83.6,43.2


Día más lluvioso:
time                   2012-10-29
temperature_2m_mean          10.1
precipitation_sum            83.6
wind_speed_10m_max           43.2
Name: 667, dtype: object


**Qué leemos:**
- 731 filas y 4 columnas, del 1 de enero de 2011 al 31 de diciembre de 2012. No falta ningún día y no hay valores nulos.
- La temperatura media diaria va de -5,2 °C a 34,0 °C, con una media de 14,7 °C: valores normales para Washington.
- En la mitad de los días no llueve nada (la mediana de la precipitación es 0), pero el máximo llega a 83,6 mm.
- El día más lluvioso es el 29 de octubre de 2012: 83,6 mm y un viento máximo de 43,2 km/h. Es el huracán Sandy.

### Conclusión: viabilidad de la API

La API de Open-Meteo sirve para el proyecto: es gratuita, no necesita clave y devuelve el tiempo real de cada día del periodo, sin huecos. El siguiente paso es cargar los datos de bicicletas y comprobar que las dos fuentes se pueden relacionar por la fecha.

### 6. Descargamos los datos de alquiler de bicicletas (UCI)

**Qué hacemos:** descargamos el zip del UCI Machine Learning Repository, lo abrimos en memoria, miramos qué archivos trae y leemos su documentación (`Readme.txt`) antes que los datos.

**Por qué descargarlo con código y no subir `day.csv` a mano:** los archivos que se suben a Colab se borran al cerrar la sesión, y quien abra el notebook no los tendría. Así el notebook funciona en cualquier sesión y en cualquier cuenta.

**Por qué `io.BytesIO`:** `requests` devuelve el zip como bytes. `BytesIO` los envuelve para que parezcan un archivo, y `zipfile` puede abrirlo sin guardarlo en disco.

**Por qué leer primero la documentación:** `season`, `weathersit` o `temp` son códigos y valores normalizados que solo se entienden con la explicación del autor.

In [ ]:
import io
import zipfile

url_uci = "https://archive.ics.uci.edu/static/public/275/bike+sharing+dataset.zip"

response_uci = requests.get(url_uci, timeout=60)
response_uci.raise_for_status()
print("Código:", response_uci.status_code, "| Tamaño:", round(len(response_uci.content) / 1024), "KB")

zip_uci = zipfile.ZipFile(io.BytesIO(response_uci.content))
zip_uci.namelist()

Código: 200 | Tamaño: 273 KB


['Readme.txt', 'day.csv', 'hour.csv']

In [ ]:
# La documentación explica cada columna y el significado de los códigos.
readme = zip_uci.read("Readme.txt").decode("utf-8", errors="replace")

inicio = readme.find("Dataset characteristics")
print(readme[inicio:inicio + 2500] if inicio != -1 else readme)

Dataset characteristics
Both hour.csv and day.csv have the following fields, except hr which is not available in day.csv
	
	- instant: record index
	- dteday : date
	- season : season (1:springer, 2:summer, 3:fall, 4:winter)
	- yr : year (0: 2011, 1:2012)
	- mnth : month ( 1 to 12)
	- hr : hour (0 to 23)
	- holiday : weather day is holiday or not (extracted from http://dchr.dc.gov/page/holiday-schedule)
	- weekday : day of the week
	- workingday : if day is neither weekend nor holiday is 1, otherwise is 0.
	+ weathersit : 
		- 1: Clear, Few clouds, Partly cloudy, Partly cloudy
		- 2: Mist + Cloudy, Mist + Broken clouds, Mist + Few clouds, Mist
		- 3: Light Snow, Light Rain + Thunderstorm + Scattered clouds, Light Rain + Scattered clouds
		- 4: Heavy Rain + Ice Pallets + Thunderstorm + Mist, Snow + Fog
	- temp : Normalized temperature in Celsius. The values are divided to 41 (max)
	- atemp: Normalized feeling temperature in Celsius. The values are divided to 50 (max)
	- hum: Normalized 

**Qué leemos:**
- El zip pesa 273 KB y trae tres archivos: `Readme.txt`, `day.csv` y `hour.csv`.
- La documentación describe cada columna. Dos puntos que comprobaremos más abajo:
  - dice `season (1:springer, 2:summer, 3:fall, 4:winter)`, es decir, que 1 es primavera (sección 12);
  - dice que `temp` está "dividida entre 41", mientras que la web de UCI da otra fórmula (sección 13).
- `casual` son los usuarios ocasionales (pase de un día o de un viaje), `registered` los socios con abono y `cnt` la suma de los dos.
- El uso del dataset exige citar a sus autores: Fanaee-T y Gama (2013). La cita va en el README.

### 7. Leemos `day.csv` y `hour.csv` y revisamos su estructura

**Qué hacemos:** leemos los dos CSV directamente de dentro del zip y revisamos cuántas filas y columnas tienen, los tipos de datos y los valores nulos.

**Por qué los dos archivos:** `day.csv` tiene una fila por día y es la que se une con el tiempo diario de Open-Meteo. `hour.csv` tiene una fila por hora y es la que necesitamos para las preguntas sobre horas punta y ventanas de mantenimiento.

**Por qué `zip_uci.open()`:** entrega cada archivo a `pd.read_csv` sin descomprimir el zip en disco.

In [ ]:
bikes_day_df = pd.read_csv(zip_uci.open("day.csv"))
bikes_hour_df = pd.read_csv(zip_uci.open("hour.csv"))

print("day.csv: ", bikes_day_df.shape)
print("hour.csv:", bikes_hour_df.shape)
bikes_day_df.head()

day.csv:  (731, 16)
hour.csv: (17379, 17)


,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,1,0,1,0,6,0,2,0.344167,0.363625,0.805833,0.160446,331,654,985
1,2,2011-01-02,1,0,1,0,0,0,2,0.363478,0.353739,0.696087,0.248539,131,670,801
2,3,2011-01-03,1,0,1,0,1,1,1,0.196364,0.189405,0.437273,0.248309,120,1229,1349
3,4,2011-01-04,1,0,1,0,2,1,1,0.200000,0.212122,0.590435,0.160296,108,1454,1562
4,5,2011-01-05,1,0,1,0,3,1,1,0.226957,0.229270,0.436957,0.186900,82,1518,1600


**Qué leemos:** `day.csv` tiene 731 filas (un día cada una) y 16 columnas; `hour.csv`, 17 379 filas y 17 columnas: las mismas más `hr`, la hora del día.

Dos años completos serían 731 × 24 = 17 544 horas, así que en `hour.csv` faltan 165. Lo estudiamos en la sección 10.

Ahora miramos los tipos de datos y los nulos de `day.csv`. Prestamos atención a `dteday`, la fecha que usaremos para relacionar los alquileres con la meteorología.

In [ ]:
print("Filas y columnas:", bikes_day_df.shape)

bikes_day_df.info()

Filas y columnas: (731, 16)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     731 non-null    int64  
 1   dteday      731 non-null    object 
 2   season      731 non-null    int64  
 3   yr          731 non-null    int64  
 4   mnth        731 non-null    int64  
 5   holiday     731 non-null    int64  
 6   weekday     731 non-null    int64  
 7   workingday  731 non-null    int64  
 8   weathersit  731 non-null    int64  
 9   temp        731 non-null    float64
 10  atemp       731 non-null    float64
 11  hum         731 non-null    float64
 12  windspeed   731 non-null    float64
 13  casual      731 non-null    int64  
 14  registered  731 non-null    int64  
 15  cnt         731 non-null    int64  
dtypes: float64(4), int64(11), object(1)
memory usage: 91.5+ KB


**Qué leemos:**
- Ninguna columna tiene valores nulos: las 16 tienen 731 valores.
- `dteday` es de tipo `object`, es decir, texto. Hay que convertirla en fecha (sección 8).
- `season`, `holiday`, `weekday`, `workingday` y `weathersit` son números enteros, pero en realidad son códigos: su significado lo comprobamos en la sección 12.
- `temp`, `atemp`, `hum` y `windspeed` son decimales entre 0 y 1 porque UCI los normalizó. Los devolvemos a sus unidades en la sección 17.
- `casual`, `registered` y `cnt` son los recuentos de alquileres: los datos que vamos a analizar.

### 8. Preparamos las fechas de ambas fuentes

**Qué hacemos:** convertimos `dteday` de los dos CSV y `time` de la API al tipo fecha con `pd.to_datetime()`.

**Por qué:** como texto, "2011-01-01" es solo una cadena de caracteres. Como fecha, pandas sabe extraer el año, el mes o el día de la semana (`.dt.year`, `.dt.dayofweek`), ordenar y comparar. Además, las dos fuentes tienen que tener el mismo tipo para poder unirlas: una fecha y un texto no coinciden aunque se escriban igual.

In [ ]:
bikes_day_df["dteday"] = pd.to_datetime(bikes_day_df["dteday"])
bikes_hour_df["dteday"] = pd.to_datetime(bikes_hour_df["dteday"])

weather_df["time"] = pd.to_datetime(weather_df["time"])

print(bikes_day_df["dteday"].dtype)
print(bikes_hour_df["dteday"].dtype)
print(weather_df["time"].dtype)

datetime64[ns]
datetime64[ns]
datetime64[ns]


**Qué leemos:** las tres columnas son ahora `datetime64[ns]`, el tipo fecha de pandas. Las dos fuentes ya se pueden comparar.

### 9. ¿Encajan las dos fuentes por la fecha?

**Qué hacemos:** convertimos las fechas de cada fuente en un conjunto (`set`) y los restamos: `a - b` devuelve los días que están en `a` y no están en `b`.

**Por qué:** la fecha es la clave que unirá las tablas en la base de datos. Si una fuente tuviera un día que la otra no tiene, un `JOIN` lo perdería sin avisar.

**Por qué no basta con comparar el número de filas:** 731 contra 731 podría esconder un día de más en un lado y uno de menos en el otro. La resta de conjuntos dice exactamente qué días sobran o faltan.

In [ ]:
print("Fechas duplicadas en bicicletas:",
      bikes_day_df["dteday"].duplicated().sum())

print("Fechas duplicadas en meteorología:",
      weather_df["time"].duplicated().sum())


fechas_bicis = set(bikes_day_df["dteday"])
fechas_tiempo = set(weather_df["time"])

print("Días en bicicletas:", len(fechas_bicis))
print("Días en meteorología:", len(fechas_tiempo))
print("Solo en bicicletas:", sorted(fechas_bicis - fechas_tiempo))
print("Solo en meteorología:", sorted(fechas_tiempo - fechas_bicis))

Fechas duplicadas en bicicletas: 0
Fechas duplicadas en meteorología: 0
Días en bicicletas: 731
Días en meteorología: 731
Solo en bicicletas: []
Solo en meteorología: []


**Qué leemos:** las dos fuentes tienen 731 días y las dos restas están vacías (`[]`): no hay ningún día que esté en una fuente y no en la otra.

No hay fechas duplicadas en ninguna de las dos tablas: cada fecha aparece una sola vez. Esto evita que se multipliquen las filas por fechas repetidas al relacionar ambas fuentes.

**Decisión:** la fecha será la clave primaria de `calendario` y de `meteorologia_dia`, y la relación entre las dos tablas será de 1 a 1.



### 10. ¿Qué horas faltan en `hour.csv`?

**Qué hacemos:** contamos cuántas horas tiene cada día con `groupby("dteday").size()`, miramos a qué horas del día faltan registros y cruzamos los días más incompletos con el tiempo de Open-Meteo.

**Por qué:** dos años completos serían 731 días × 24 horas = 17 544 filas. Si faltan justamente las horas sin alquileres, la media por hora se calcula solo con las horas que sí tuvieron y sale inflada. Hay que saber cuáles faltan y por qué.

**Por qué `size()` y no `count()`:** `size()` cuenta filas; `count()` cuenta los valores no nulos de cada columna. Para saber cuántas horas hay por día, lo que importa son las filas.

**Por qué cruzar con el tiempo:** si los días incompletos coinciden con temporales, lo más probable es que el sistema cerrara, y eso es información útil para el negocio.

In [ ]:
horas_por_dia = bikes_hour_df.groupby("dteday").size()

print("Filas en hour.csv:", len(bikes_hour_df), "de", 731 * 24)
print("Días con las 24 horas:", (horas_por_dia == 24).sum())
print("Días con menos de 24 horas:", (horas_por_dia < 24).sum())
horas_por_dia.sort_values().head(10)

Filas en hour.csv: 17379 de 17544
Días con las 24 horas: 655
Días con menos de 24 horas: 76


,0
dteday,
2012-10-29,1
2011-01-27,8
2012-10-30,11
2011-01-18,12
2011-01-26,16
2011-08-28,17
2011-08-27,18
2011-02-22,18
2011-02-28,22


In [ ]:
# ¿A qué horas del día faltan registros? (731 días menos los días en que aparece esa hora)
dias_sin_esa_hora = len(bikes_day_df) - bikes_hour_df.groupby("hr").size()
dias_sin_esa_hora.sort_values(ascending=False).head(8)

,0
hr,
3,34
4,34
2,16
5,14
1,7
6,6
0,5
7,4


In [ ]:
# ¿Coinciden los días con muchas horas perdidas con mal tiempo?
dias_incompletos = horas_por_dia[horas_por_dia < 20].rename("horas").reset_index()
dias_incompletos.merge(weather_df, left_on="dteday", right_on="time")[
    ["dteday", "horas", "precipitation_sum", "wind_speed_10m_max", "temperature_2m_mean"]
]

,dteday,horas,precipitation_sum,wind_speed_10m_max,temperature_2m_mean
0,2011-01-18,12,4.9,14.7,0.6
1,2011-01-26,16,24.7,23.7,0.7
2,2011-01-27,8,0.6,21.3,-0.6
3,2011-02-22,18,5.7,15.1,-0.8
4,2011-08-27,18,54.1,36.7,23.3
5,2011-08-28,17,48.1,39.7,23.3
6,2012-10-29,1,83.6,43.2,10.1
7,2012-10-30,11,31.3,35.6,5.8


**Qué leemos:** `hour.csv` tiene 17 379 filas de las 17 544 posibles: faltan 165 horas. 655 días están completos y 76 tienen alguna hora de menos.

- Las ausencias se concentran principalmente en la madrugada. Las 03:00 y las 04:00 no aparecen en 34 días cada una. Estos datos no permiten confirmar si se debe a ausencia de alquileres, cierres del servicio o problemas de registro.
- Los días con más horas perdidas coinciden con temporales: el huracán Irene (27 y 28 de agosto de 2011, más de 48 mm de lluvia cada día), el huracán Sandy (29 y 30 de octubre de 2012, con 83,6 mm el día 29 y una sola hora registrada) y un temporal de invierno el 26 y 27 de enero de 2011.

**Decisión:** no rellenamos las horas que faltan con ceros: no sabemos si no hubo alquileres o si el sistema estaba cerrado. Lo explicamos en las limitaciones del README.

### 11. ¿Cuadran los totales?

**Qué hacemos:** comprobamos que `casual + registered = cnt` en los dos archivos, y que la suma de las horas de cada día da el total de `day.csv`.

**Por qué:** si los totales cuadran, `day.csv` no aporta ningún recuento que no esté ya en `hour.csv`. En la base de datos basta con guardar los alquileres por hora; los totales por día se calculan con `GROUP BY`. Guardar los dos sería tener el mismo dato en dos sitios, con el riesgo de que un día dejen de coincidir.

**Por qué `set_index("dteday")`:** para que la resta de la última línea compare cada día con el mismo día, alineando por fecha y no por posición.

In [ ]:
print("day.csv, filas donde casual + registered ≠ cnt:",
      (bikes_day_df["casual"] + bikes_day_df["registered"] != bikes_day_df["cnt"]).sum())
print("hour.csv, filas donde casual + registered ≠ cnt:",
      (bikes_hour_df["casual"] + bikes_hour_df["registered"] != bikes_hour_df["cnt"]).sum())

suma_horas = bikes_hour_df.groupby("dteday")["cnt"].sum()
total_dia = bikes_day_df.set_index("dteday")["cnt"]
print("Días donde la suma de las horas ≠ day.csv:", (suma_horas != total_dia).sum())

day.csv, filas donde casual + registered ≠ cnt: 0
hour.csv, filas donde casual + registered ≠ cnt: 0
Días donde la suma de las horas ≠ day.csv: 0


**Qué leemos:** Los totales coinciden en ambas tablas: los alquileres ocasionales y registrados suman el total, y la suma de los registros horarios coincide con el total diario.

Por ello, podemos calcular los alquileres diarios a partir de la tabla horaria, sin almacenar los mismos totales dos veces en la base de datos.

Esta comprobación confirma la coherencia entre las tablas, aunque no explica por qué faltan algunas horas.

**Decisión:** en la base de datos guardamos solo la tabla por horas; los totales por día salen con `GROUP BY` en la vista `v_dia` (sección 20).

### 12. ¿Qué significan los códigos?

**Qué hacemos:** comprobamos con las fechas qué significa cada código, en lugar de fiarnos solo de la documentación.

- `season`: una tabla cruzada (`pd.crosstab`) cuenta cuántos días de cada mes tienen cada código. Si 1 es invierno, tiene que aparecer en enero y febrero; si es primavera, en abril y mayo.
- `weekday`: UCI cuenta los días desde el domingo (0 = domingo) y pandas desde el lunes (0 = lunes). `(x + 1) % 7` pasa del sistema de pandas al de UCI: el lunes pasa de 0 a 1 y el domingo, de 6 a 0.
- `weathersit`: contamos cuántas horas hay de cada situación con `value_counts()`.

**Por qué:** en la sección 6 vimos que el `Readme.txt` dice 1 = primavera. Si usáramos esa traducción sin comprobarla, todos los gráficos por estación del año estarían mal.

In [ ]:
# Filas: mes. Columnas: código de estación. Cada celda: número de días.
pd.crosstab(bikes_day_df["mnth"], bikes_day_df["season"])

season,1,2,3,4
mnth,,,,
1,62,0,0,0
2,57,0,0,0
3,40,22,0,0
4,0,60,0,0
5,0,62,0,0
6,0,40,20,0
7,0,0,62,0
8,0,0,62,0
9,0,0,44,16


In [ ]:
weekday_desde_fecha = (bikes_day_df["dteday"].dt.dayofweek + 1) % 7
print("Días donde weekday no coincide con la fecha:", (weekday_desde_fecha != bikes_day_df["weekday"]).sum())

# Horas de cada situación meteorológica (1 = despejado ... 4 = lluvia fuerte)
bikes_hour_df["weathersit"].value_counts().sort_index()

Días donde weekday no coincide con la fecha: 0


,count
weathersit,
1,11413
2,4544
3,1419
4,3


**Qué leemos:**

- **Estaciones (`season`):** al comparar los códigos con los meses, comprobamos que 1 = invierno, 2 = primavera, 3 = verano y 4 = otoño. Esta correspondencia no coincide con las etiquetas del `Readme.txt`, por lo que corregimos su interpretación.

- Marzo, junio, septiembre y diciembre contienen dos estaciones porque el cambio ocurre dentro del mes. Los números de la tabla representan días registrados entre 2011 y 2012, no alquileres.

- **Día de la semana (`weekday`):** coincide con la fecha en los 731 días. Para comprobarlo, adaptamos la numeración: UCI empieza en domingo y pandas en lunes.

- **Situación meteorológica (`weathersit`):** encontramos 11.413 registros horarios en la categoría 1, 4.544 en la 2, 1.419 en la 3 y solo 3 en la 4. La categoría 1 es la más frecuente y la 4 tiene muy pocos registros para analizarla por separado.

**Decisiones:**

- Usamos los nombres correctos de las estaciones al preparar la tabla `calendario`.
- Guardamos los códigos meteorológicos y sus descripciones en la tabla `situacion_meteo`.
- Conservamos las cuatro categorías originales. Para las comparaciones generales del análisis, agruparemos las categorías 3 y 4 como condiciones de lluvia o nieve, indicando que esta agrupación combina distintas intensidades.

### 13. ¿Coinciden las temperaturas de las dos fuentes?

**Qué hacemos:** convertimos la temperatura normalizada de UCI (`temp`) con las dos fórmulas que da su documentación y comparamos cada una con la temperatura media de Open-Meteo del mismo día.

**Por qué:** la documentación se contradice. El `Readme.txt` dice que la temperatura se dividió entre 41; la web de UCI dice que se normalizó con un mínimo de -8 °C y un máximo de 39 °C, es decir, `(t + 8) / 47`. Para volver a grados, una fórmula multiplica por 41 y la otra multiplica por 47 y resta 8. Utilizamos la temperatura de Open-Meteo como referencia para comparar las dos conversiones.

**Qué medimos:**
- la **correlación**: si las dos fuentes suben y bajan a la vez (cerca de 1), describen los mismos días en el mismo sitio;
- la **diferencia media**: si una fórmula da siempre más calor o más frío que la realidad;
- el **error medio absoluto**: cuántos grados se equivoca de media cada fórmula, sin que los errores positivos y negativos se compensen.

In [ ]:
comparacion = bikes_day_df[["dteday", "temp"]].merge(weather_df, left_on="dteday", right_on="time")

print("Correlación entre temp de UCI y temperatura de Open-Meteo:",
      round(comparacion["temp"].corr(comparacion["temperature_2m_mean"]), 3))
print()

opciones = {
    "temp × 41 (Readme del zip)": comparacion["temp"] * 41,
    "temp × 47 - 8 (web de UCI)": comparacion["temp"] * 47 - 8,
}
for nombre, estimada in opciones.items():
    diferencia = estimada - comparacion["temperature_2m_mean"]
    print(f"{nombre}: diferencia media {diferencia.mean():+.1f} °C | error medio absoluto {diferencia.abs().mean():.1f} °C")

Correlación entre temp de UCI y temperatura de Open-Meteo: 0.996

temp × 41 (Readme del zip): diferencia media +5.6 °C | error medio absoluto 5.6 °C
temp × 47 - 8 (web de UCI): diferencia media +0.6 °C | error medio absoluto 0.9 °C


**Qué leemos:**

Las temperaturas de UCI y Open-Meteo evolucionan de forma muy similar: su correlación es de 0,996.

Al convertir los valores de UCI a grados Celsius, la fórmula `temp * 41` produce temperaturas 5,6 °C superiores a Open-Meteo de media.

Con `temp * 47 - 8`, la diferencia media es de +0,6 °C y la diferencia absoluta media es de 0,9 °C. Esta segunda medida cuenta todas las diferencias como positivas para evitar que se compensen.

**Decisión:**

Usamos `temp * 47 - 8`, la fórmula indicada en la web de UCI, para recuperar la temperatura en grados Celsius. La comparación con Open-Meteo respalda esta elección.

Las dos fuentes no tienen por qué dar valores idénticos. Utilizamos Open-Meteo como referencia de comparación, no como una medida exacta de la temperatura real.

### Conclusión de la parte 1: las dos fuentes se pueden unir

- **Fechas:** ambas fuentes contienen los mismos 731 días, sin fechas duplicadas. Podemos relacionarlas mediante la fecha.
- **Horas:** faltan 165 registros horarios. Conservamos las horas disponibles sin rellenar las ausencias con ceros, porque desconocemos su causa.
- **Totales:** la suma de los alquileres horarios coincide con el total diario. Podemos guardar los alquileres por hora y calcular los totales diarios cuando los necesitemos.
- **Códigos:** comprobamos el significado de las estaciones y la correspondencia del día de la semana con la fecha. La categoría meteorológica 4 solo tiene tres registros horarios.
- **Temperatura:** usamos `temp * 47 - 8` para convertir los valores de UCI a grados Celsius. Esta conversión presenta una diferencia absoluta media de 0,9 °C respecto a Open-Meteo.

**Siguiente paso:** preparar las tablas limpias, definir sus relaciones y cargarlas en MySQL.

## Parte 2 · Limpieza y base de datos MySQL

**Qué hacemos:** con las dos fuentes revisadas, construimos en pandas las cuatro tablas del modelo (secciones 14 a 17), comprobamos sus claves (18), instalamos MySQL en Colab (19), creamos el esquema (20), cargamos los datos (21), probamos la base (22) y guardamos todo en Google Drive (23).

**El modelo:**

| Tabla | Una fila por | Clave primaria | Claves foráneas |
|---|---|---|---|
| `situacion_meteo` | código de tiempo de UCI (1-4) | `situacion_id` | |
| `calendario` | día | `fecha` | |
| `meteorologia_dia` | día | `fecha` | `fecha` → `calendario` |
| `alquileres_hora` | día y hora | `fecha` + `hora` | `fecha` → `calendario`, `situacion_id` → `situacion_meteo` |

`alquileres_hora` es la tabla de hechos: guarda lo que medimos, los alquileres. `calendario`, `meteorologia_dia` y `situacion_meteo` la describen: cuándo, con qué tiempo y qué significa cada código. `calendario` y `meteorologia_dia` se relacionan 1 a 1 por la fecha.

**Por qué no una sola tabla grande:** repetiríamos el día de la semana, la estación y el tiempo del día en cada una de sus 24 horas. Separarlas evita guardar el mismo dato muchas veces, y es lo que pide el proyecto: al menos tres tablas unidas por claves.

### 14. Tabla `situacion_meteo`: qué significa cada código de tiempo

**Qué hacemos:** creamos una tabla pequeña con los cuatro códigos de `weathersit` y su descripción en español.

**Por qué una tabla y no el texto en cada fila:** en `alquileres_hora` el código se repite 17 379 veces. Guardar ahí el número, y su significado una sola vez en otra tabla, ocupa menos, evita erratas y permite que la base rechace un código que no existe (clave foránea).

**De dónde salen los textos:** traducen las categorías de la documentación de UCI, resumidas para que se lean bien en un gráfico.

In [ ]:
situacion_meteo = pd.DataFrame({
    "situacion_id": [1, 2, 3, 4],
    "descripcion": [
        "Despejado o poco nuboso",
        "Niebla o nublado",
        "Lluvia o nieve ligera, tormenta",
        "Lluvia fuerte, granizo o nieve",
    ],
})
situacion_meteo

,situacion_id,descripcion
0,1,Despejado o poco nuboso
1,2,Niebla o nublado
2,3,"Lluvia o nieve ligera, tormenta"
3,4,"Lluvia fuerte, granizo o nieve"


**Qué leemos:**

Creamos una tabla con los cuatro códigos meteorológicos y sus descripciones. Cada código aparece una sola vez.

**Decisión:**

`situacion_id` será la clave primaria de esta tabla. En `alquileres_hora`, ese mismo campo será una clave foránea que nos permitirá consultar la descripción de las condiciones meteorológicas de cada registro.

### 15. Tabla `calendario`: una fila por día

**Qué hacemos:** a partir de `day.csv` creamos una tabla con la fecha y todo lo que describe ese día: año, mes, día de la semana, si es fin de semana, festivo o laborable, y la estación del año.

**Qué cambiamos respecto a UCI:**
- el día de la semana lo calculamos desde la fecha, con 1 = lunes, porque el `weekday` de UCI empieza en domingo = 0 y confunde al leer las consultas;
- añadimos `nombre_dia` y `es_fin_de_semana`, que UCI no trae (solo trae `workingday`, que mezcla fines de semana y festivos);
- traducimos `season` con lo comprobado en la sección 12 (1 = invierno);
- los nombres de las columnas pasan al español, sin abreviaturas.

**Por qué la fecha como texto `AAAA-MM-DD`:** es el formato que MySQL convierte directamente en `DATE` al cargar.

**Por qué `dict(enumerate(NOMBRES_DIA))`:** `enumerate` numera la lista desde 0 (0 lunes, 1 martes...), igual que `dayofweek` de pandas; `map` cambia cada número por su nombre.

In [ ]:
NOMBRES_DIA = ["lunes", "martes", "miércoles", "jueves", "viernes", "sábado", "domingo"]
ESTACIONES = {1: "invierno", 2: "primavera", 3: "verano", 4: "otoño"}   # comprobado en la sección 12

fecha = bikes_day_df["dteday"]
calendario = pd.DataFrame({
    "fecha": fecha.dt.strftime("%Y-%m-%d"),
    "anio": fecha.dt.year,
    "mes": fecha.dt.month,
    "dia_semana": fecha.dt.dayofweek + 1,                     # 1 = lunes ... 7 = domingo
    "nombre_dia": fecha.dt.dayofweek.map(dict(enumerate(NOMBRES_DIA))),
    "es_fin_de_semana": (fecha.dt.dayofweek >= 5).astype(int),
    "es_festivo": bikes_day_df["holiday"],
    "es_laborable": bikes_day_df["workingday"],
    "estacion": bikes_day_df["season"].map(ESTACIONES),
})
print(calendario.shape)
calendario.head()

(731, 9)


,fecha,anio,mes,dia_semana,nombre_dia,es_fin_de_semana,es_festivo,es_laborable,estacion
0,2011-01-01,2011,1,6,sábado,1,0,0,invierno
1,2011-01-02,2011,1,7,domingo,1,0,0,invierno
2,2011-01-03,2011,1,1,lunes,0,0,1,invierno
3,2011-01-04,2011,1,2,martes,0,0,1,invierno
4,2011-01-05,2011,1,3,miércoles,0,0,1,invierno


**Qué leemos:**

Preparamos una tabla con una fila por día y sus características: año, mes, día de la semana, fin de semana, festivo, laborable y estación.

Numeramos los días desde 1 (lunes) hasta 7 (domingo). Los indicadores usan 1 para sí y 0 para no. Conservamos los valores de festivo y laborable del dataset original.

**Decisión:**

La fecha será la clave primaria de `calendario`. Esta tabla permitirá agrupar los alquileres por mes, estación o tipo de día sin repetir toda esta información en cada registro horario.

### 16. Tabla `meteorologia_dia`: el tiempo de Open-Meteo

**Qué hacemos:** cambiamos los nombres de las columnas al español, con la unidad en el nombre (`_c`, `_mm`, `_kmh`), y añadimos `dia_lluvioso`.

**Por qué la unidad en el nombre:** quien lea una consulta sabe en qué unidad está cada valor sin buscar la documentación.

**Por qué el umbral de 1 mm:** es la convención meteorológica para "día de lluvia". Por debajo es llovizna o ruido del modelo, que no cambia el comportamiento de nadie. Lo guardamos en una variable (`UMBRAL_LLUVIA_MM`) para poder cambiarlo en un solo sitio.

**Por qué `astype(int)`:** la comparación da `True` o `False`; en la base de datos lo guardamos como 1 y 0, que se pueden sumar para contar días de lluvia.

In [ ]:
UMBRAL_LLUVIA_MM = 1.0

meteorologia_dia = pd.DataFrame({
    "fecha": weather_df["time"].dt.strftime("%Y-%m-%d"),
    "temp_media_c": weather_df["temperature_2m_mean"],
    "precipitacion_mm": weather_df["precipitation_sum"],
    "viento_max_kmh": weather_df["wind_speed_10m_max"],
    "dia_lluvioso": (weather_df["precipitation_sum"] >= UMBRAL_LLUVIA_MM).astype(int),
})
print(meteorologia_dia.shape)
print("Días de lluvia:", meteorologia_dia["dia_lluvioso"].sum(), "de", len(meteorologia_dia))
meteorologia_dia.head()

(731, 5)
Días de lluvia: 241 de 731


,fecha,temp_media_c,precipitacion_mm,viento_max_kmh,dia_lluvioso
0,2011-01-01,8.2,2.5,17.7,1
1,2011-01-02,8.6,3.0,18.9,1
2,2011-01-03,-0.1,0.0,17.4,0
3,2011-01-04,0.1,0.0,11.2,0
4,2011-01-05,-0.6,0.0,13.9,0


**Qué leemos:**

Las primeras filas conservan los valores de Open-Meteo. Los días 1 y 2 de enero tienen 2,5 y 3 mm de precipitación, por lo que reciben un 1 en `dia_lluvioso`. Los tres días siguientes tienen 0 mm y reciben un 0.

Según el recuento anotado en el notebook, 241 de los 731 días, aproximadamente un 33 %, alcanzan el umbral de 1 mm.

**Decisión:**

Utilizaremos esta clasificación para comparar los alquileres entre días con precipitación de al menos 1 mm y días por debajo de ese umbral.

### 17. Tabla `alquileres_hora`: los alquileres de cada hora

**Qué hacemos:** a partir de `hour.csv` creamos la tabla de hechos: una fila por día y hora, con los alquileres de ocasionales y registrados y el tiempo de esa hora.

**Qué cambiamos respecto a UCI:**
- devolvemos el tiempo a sus unidades con la fórmula comprobada en la sección 13, `valor × (máximo - mínimo) + mínimo`: -8 y 39 °C para la temperatura, -16 y 50 °C para la sensación térmica;
- la humedad se multiplica por 100 para tenerla en %;
- el viento se multiplica por 67, como indica UCI, pero su documentación no dice la unidad: lo llamamos simplemente `viento`;
- `casual` pasa a `ocasionales`, `registered` a `registrados` y `cnt` a `total`.

**Por qué no guardamos `season`, `weekday`, `holiday` ni `workingday`:** ya están en `calendario` y se recuperan con un `JOIN` por la fecha. Repetirlos en cada hora sería guardar el mismo dato 24 veces.

**Por qué los valores en variables (`TEMP_MIN`, `TEMP_MAX`...):** para que se vea de dónde sale cada número y se pueda cambiar en un solo sitio.

In [ ]:
# Valores de la web de UCI para deshacer la normalización (comprobados en la sección 13)
TEMP_MIN, TEMP_MAX = -8, 39
ATEMP_MIN, ATEMP_MAX = -16, 50

alquileres_hora = pd.DataFrame({
    "fecha": bikes_hour_df["dteday"].dt.strftime("%Y-%m-%d"),
    "hora": bikes_hour_df["hr"],
    "situacion_id": bikes_hour_df["weathersit"],
    "temp_c": (bikes_hour_df["temp"] * (TEMP_MAX - TEMP_MIN) + TEMP_MIN).round(1),
    "sensacion_c": (bikes_hour_df["atemp"] * (ATEMP_MAX - ATEMP_MIN) + ATEMP_MIN).round(1),
    "humedad_pct": (bikes_hour_df["hum"] * 100).round().astype(int),
    "viento": (bikes_hour_df["windspeed"] * 67).round(1),        # UCI no indica la unidad
    "ocasionales": bikes_hour_df["casual"],
    "registrados": bikes_hour_df["registered"],
    "total": bikes_hour_df["cnt"],
})
print(alquileres_hora.shape)
alquileres_hora.head()

(17379, 10)


,fecha,hora,situacion_id,temp_c,sensacion_c,humedad_pct,viento,ocasionales,registrados,total
0,2011-01-01,0,1,3.3,3.0,81,0.0,3,13,16
1,2011-01-01,1,1,2.3,2.0,80,0.0,8,32,40
2,2011-01-01,2,1,2.3,2.0,80,0.0,5,27,32
3,2011-01-01,3,1,3.3,3.0,75,0.0,3,10,13
4,2011-01-01,4,1,3.3,3.0,75,0.0,0,1,1


**Qué leemos:**

La tabla contiene 17.379 filas y 10 columnas. Conservamos los registros horarios disponibles del archivo original.

El primer registro corresponde a las 00:00 del 1 de enero de 2011: tiene una temperatura de 3,3 °C, una humedad del 81 % y 16 alquileres, repartidos entre 3 ocasionales y 13 registrados.

**Decisión:**

Utilizaremos esta tabla para analizar cómo varían los alquileres según la hora y el tipo de usuario. También podremos sumar sus registros para obtener totales diarios.

Los recuentos representan alquileres, no personas diferentes.

### 18. Comprobamos las claves antes de cargar

**Qué hacemos:** revisamos en pandas las mismas reglas que después impondrá MySQL: claves primarias sin repetir, claves foráneas que apuntan a filas que existen y ninguna estación sin traducir.

**Por qué antes de cargar:** si algo falla dentro de MySQL, el error solo dice que no se cumple una restricción ("a foreign key constraint fails"), sin decir cuántas filas ni cuáles. Aquí vemos cuántas y en qué tabla, y se corrige antes.

**Por qué `duplicated(["fecha", "hora"])`:** la clave primaria de `alquileres_hora` son las dos columnas juntas. Una fecha se repite 24 veces y una hora 731 veces; lo que no puede repetirse es la pareja.

**Por qué `~` e `isin`:** `isin` marca con `True` las filas cuyo valor está en la lista; `~` lo invierte, así que se cuentan las que **no** están.

In [ ]:
fechas = set(calendario["fecha"])

comprobaciones = {
    "calendario: fecha repetida": calendario["fecha"].duplicated().sum(),
    "meteorologia_dia: fecha repetida": meteorologia_dia["fecha"].duplicated().sum(),
    "alquileres_hora: fecha y hora repetidas": alquileres_hora.duplicated(["fecha", "hora"]).sum(),
    "meteorologia_dia: fecha que no está en calendario": (~meteorologia_dia["fecha"].isin(fechas)).sum(),
    "alquileres_hora: fecha que no está en calendario": (~alquileres_hora["fecha"].isin(fechas)).sum(),
    "alquileres_hora: código de tiempo desconocido": (~alquileres_hora["situacion_id"].isin(situacion_meteo["situacion_id"])).sum(),
    "calendario: estación sin traducir": calendario["estacion"].isna().sum(),
}
pd.Series(comprobaciones, name="problemas")

,problemas
calendario: fecha repetida,0
meteorologia_dia: fecha repetida,0
alquileres_hora: fecha y hora repetidas,0
meteorologia_dia: fecha que no está en calendario,0
alquileres_hora: fecha que no está en calendario,0
alquileres_hora: código de tiempo desconocido,0
calendario: estación sin traducir,0


**Qué leemos:** 0 problemas en las siete comprobaciones. La carga en MySQL no fallará por claves repetidas ni por claves foráneas que apunten a filas inexistentes.

### 19. Instalamos MySQL dentro de Colab

**Qué hacemos:** instalamos el servidor MySQL en la máquina de Colab, lo arrancamos, instalamos el conector de Python (`pymysql`) y creamos la base de datos `bikeshare` con un usuario propio.

**Por qué MySQL:** es el sistema de bases de datos que pide el proyecto. Es un servidor: guarda los datos en tablas relacionadas y responde a consultas SQL, que es lo que haremos en el notebook de informe.

**Por qué dentro de Colab y no en el portátil:** así todo el equipo ejecuta el mismo notebook sin instalar nada. La máquina de Colab se borra al cerrar la sesión, por eso esta celda instala MySQL cada vez (1-2 minutos). Lo que se conserva es el volcado `.sql` que guardamos en Drive en la sección 23: con él, la base se abre en MySQL Workbench.

**Qué hace cada comando:**
- `apt-get install mysql-server`: instala el servidor; `> /dev/null` oculta los cientos de líneas de la instalación.
- `service mysql start`: lo pone en marcha.
- `mysql -e "..."`: ejecuta una instrucción SQL desde la terminal. En Colab somos el administrador (`root`) y entramos sin contraseña.
- Las llaves de `{DB_NAME}` hacen que Colab ponga ahí el valor de la variable de Python.

**Por qué un usuario propio y no `root`:** `root` solo puede entrar desde la terminal de la máquina. pandas se conecta como un programa externo y necesita un usuario con contraseña. Esta contraseña solo existe dentro de la máquina de Colab y desaparece con ella.

**Por qué `utf8mb4`:** es la codificación que acepta cualquier carácter, incluidas las tildes y la ñ de "otoño".

In [ ]:
# 1) Instalamos el servidor MySQL en la máquina de Colab (1-2 minutos en cada sesión).
!apt-get -qq update > /dev/null
!DEBIAN_FRONTEND=noninteractive apt-get -qq install -y mysql-server > /dev/null
!service mysql start

# 2) Instalamos el conector de Python para MySQL.
!pip -q install pymysql

# 3) Creamos la base de datos y un usuario propio para pandas.
DB_NAME = "bikeshare"
DB_USER = "bikeshare"
DB_PASSWORD = "bikeshare_colab"   # solo existe dentro de esta máquina de Colab

!mysql -e "CREATE DATABASE IF NOT EXISTS {DB_NAME} CHARACTER SET utf8mb4;"
!mysql -e "CREATE USER IF NOT EXISTS '{DB_USER}'@'localhost' IDENTIFIED BY '{DB_PASSWORD}';"
!mysql -e "GRANT ALL PRIVILEGES ON {DB_NAME}.* TO '{DB_USER}'@'localhost'; FLUSH PRIVILEGES;"
!mysql -e "SELECT VERSION() AS version_mysql;"

W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu noble InRelease' does not seem to provide it (sources.list entry misspelt?)
 * Starting MySQL database server mysqld        su: warning: cannot change directory to /nonexistent: No such file or directory
[ OK ]
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.7/46.7 kB 1.4 MB/s eta 0:00:00
+-------------------------+
| version_mysql           |
+-------------------------+
| 8.0.46-0ubuntu0.24.04.4 |
+-------------------------+


**Qué leemos:** MySQL 8.0 instalado y en marcha (`[ OK ]`), con la base `bikeshare` y el usuario creados. Los avisos que empiezan por `W:` y `su: warning` son normales en Colab y no afectan a nada.

### 20. Creamos el esquema: tablas, claves y vista

**Qué hacemos:** escribimos el esquema de la base de datos, lo guardamos en `01_schema.sql` y lo ejecutamos con el cliente de MySQL, igual que se haría en MySQL Workbench.

**Por qué ejecutarlo desde un archivo `.sql`:** el esquema en archivo es uno de los entregables. Ejecutándolo desde el archivo comprobamos que el archivo entregado funciona.

**Cómo está hecho:**
- **Primero se borra lo que hubiera** (`DROP ... IF EXISTS`), empezando por lo que depende de otras tablas: la vista, después `alquileres_hora` y al final las tablas a las que apuntan las claves. Así el esquema se puede ejecutar varias veces.
- **Claves primarias** (`PRIMARY KEY`): identifican cada fila. En `alquileres_hora` es la pareja `fecha` + `hora`.
- **Claves foráneas** (`FOREIGN KEY ... REFERENCES`): MySQL rechaza un alquiler cuya fecha no esté en `calendario` o cuyo código de tiempo no esté en `situacion_meteo`.
- **`CHECK`**: reglas que MySQL comprueba en cada fila, como que la hora esté entre 0 y 23 o que `total = ocasionales + registrados`.
- **`ENGINE = InnoDB`**: el motor de MySQL que respeta las claves foráneas. Es el de defecto en MySQL 8; lo escribimos para que quede claro.
- **Tipos ajustados al dato:** `TINYINT` para números pequeños (horas, meses, 0/1), `SMALLINT` para los recuentos, `DECIMAL(4,1)` para valores con un decimal y `DATE` para las fechas.

**La vista `v_dia`:** junta las tres tablas y suma los alquileres de cada día. Muchas preguntas son por día; la vista evita repetir el mismo `JOIN` y `GROUP BY` en cada consulta.

In [ ]:
ESQUEMA = """
-- Capital Bikeshare y el tiempo en Washington D. C. (2011-2012). Esquema para MySQL 8.
-- Se borra primero lo que depende de otras tablas, para poder ejecutarlo varias veces.

DROP VIEW IF EXISTS v_dia;
DROP TABLE IF EXISTS alquileres_hora;
DROP TABLE IF EXISTS meteorologia_dia;
DROP TABLE IF EXISTS calendario;
DROP TABLE IF EXISTS situacion_meteo;

CREATE TABLE situacion_meteo (
    situacion_id  TINYINT      NOT NULL,
    descripcion   VARCHAR(60)  NOT NULL,
    PRIMARY KEY (situacion_id)
) ENGINE = InnoDB;

CREATE TABLE calendario (
    fecha             DATE         NOT NULL,
    anio              SMALLINT     NOT NULL,
    mes               TINYINT      NOT NULL,
    dia_semana        TINYINT      NOT NULL,   -- 1 = lunes ... 7 = domingo
    nombre_dia        VARCHAR(10)  NOT NULL,
    es_fin_de_semana  TINYINT      NOT NULL,   -- 0 / 1
    es_festivo        TINYINT      NOT NULL,   -- 0 / 1
    es_laborable      TINYINT      NOT NULL,   -- 1 si no es fin de semana ni festivo
    estacion          VARCHAR(10)  NOT NULL,
    PRIMARY KEY (fecha),
    CHECK (dia_semana BETWEEN 1 AND 7)
) ENGINE = InnoDB;

CREATE TABLE meteorologia_dia (
    fecha             DATE          NOT NULL,
    temp_media_c      DECIMAL(4,1),
    precipitacion_mm  DECIMAL(5,1),
    viento_max_kmh    DECIMAL(5,1),
    dia_lluvioso      TINYINT       NOT NULL,  -- 1 si precipitacion_mm >= 1
    PRIMARY KEY (fecha),
    FOREIGN KEY (fecha) REFERENCES calendario (fecha)
) ENGINE = InnoDB;

CREATE TABLE alquileres_hora (
    fecha          DATE          NOT NULL,
    hora           TINYINT       NOT NULL,     -- 0 a 23, hora local de Washington
    situacion_id   TINYINT       NOT NULL,
    temp_c         DECIMAL(4,1),
    sensacion_c    DECIMAL(4,1),
    humedad_pct    TINYINT,
    viento         DECIMAL(4,1),
    ocasionales    SMALLINT      NOT NULL,
    registrados    SMALLINT      NOT NULL,
    total          SMALLINT      NOT NULL,
    PRIMARY KEY (fecha, hora),
    FOREIGN KEY (fecha) REFERENCES calendario (fecha),
    FOREIGN KEY (situacion_id) REFERENCES situacion_meteo (situacion_id),
    CHECK (hora BETWEEN 0 AND 23),
    CHECK (total = ocasionales + registrados)
) ENGINE = InnoDB;

CREATE VIEW v_dia AS
SELECT
    c.fecha, c.anio, c.mes, c.dia_semana, c.nombre_dia,
    c.es_fin_de_semana, c.es_festivo, c.es_laborable, c.estacion,
    m.temp_media_c, m.precipitacion_mm, m.viento_max_kmh, m.dia_lluvioso,
    SUM(a.ocasionales) AS ocasionales,
    SUM(a.registrados) AS registrados,
    SUM(a.total)       AS total
FROM calendario AS c
JOIN meteorologia_dia AS m ON m.fecha = c.fecha
JOIN alquileres_hora  AS a ON a.fecha = c.fecha
GROUP BY
    c.fecha, c.anio, c.mes, c.dia_semana, c.nombre_dia,
    c.es_fin_de_semana, c.es_festivo, c.es_laborable, c.estacion,
    m.temp_media_c, m.precipitacion_mm, m.viento_max_kmh, m.dia_lluvioso;
"""

with open("01_schema.sql", "w", encoding="utf-8") as f:
    f.write(ESQUEMA)

!mysql {DB_NAME} < 01_schema.sql
!mysql {DB_NAME} -e "SHOW FULL TABLES;"

+---------------------+------------+
| Tables_in_bikeshare | Table_type |
+---------------------+------------+
| alquileres_hora     | BASE TABLE |
| calendario          | BASE TABLE |
| meteorologia_dia    | BASE TABLE |
| situacion_meteo     | BASE TABLE |
| v_dia               | VIEW       |
+---------------------+------------+


**Qué leemos:** MySQL ha creado las cuatro tablas (`BASE TABLE`) y la vista `v_dia` (`VIEW`). El archivo `01_schema.sql` funciona.

### 21. Cargamos las tablas en MySQL

**Qué hacemos:** conectamos pandas con MySQL y cargamos las cuatro tablas en orden.

**Cómo se conecta pandas a MySQL:** con un *engine* de SQLAlchemy, que recibe una dirección con todo lo necesario: `mysql+pymysql://usuario:contraseña@servidor/base`. `pymysql` es el conector que habla con MySQL.

**Por qué en este orden:** primero las tablas a las que apuntan las claves foráneas (`situacion_meteo` y `calendario`) y después las que dependen de ellas. Al revés, MySQL rechazaría los alquileres porque sus fechas todavía no existirían en `calendario`.

**Por qué `if_exists="append"` y no `"replace"`:** las tablas ya existen, creadas por el esquema con sus claves. `append` añade las filas a esas tablas; `replace` las borraría y crearía otras nuevas sin claves primarias ni foráneas.

**Por qué `chunksize=1000`:** envía las 17 379 filas de `alquileres_hora` en grupos de 1000 en lugar de todas de golpe.

**Ojo:** si hay que repetir la carga, hay que ejecutar antes la sección 20. Si no, las filas se duplican y la clave primaria da error.

In [ ]:
from sqlalchemy import create_engine

engine = create_engine(f"mysql+pymysql://{DB_USER}:{DB_PASSWORD}@localhost/{DB_NAME}?charset=utf8mb4")

tablas = [
    ("situacion_meteo", situacion_meteo),
    ("calendario", calendario),
    ("meteorologia_dia", meteorologia_dia),
    ("alquileres_hora", alquileres_hora),
]
for nombre, tabla in tablas:
    tabla.to_sql(nombre, engine, if_exists="append", index=False, chunksize=1000)

pd.read_sql_query("""
    SELECT 'situacion_meteo' AS tabla, COUNT(*) AS filas FROM situacion_meteo
    UNION ALL SELECT 'calendario',       COUNT(*) FROM calendario
    UNION ALL SELECT 'meteorologia_dia', COUNT(*) FROM meteorologia_dia
    UNION ALL SELECT 'alquileres_hora',  COUNT(*) FROM alquileres_hora
""", engine)

,tabla,filas
0,situacion_meteo,4
1,calendario,731
2,meteorologia_dia,731
3,alquileres_hora,17379


**Qué leemos:** las filas de cada tabla en MySQL coinciden con las de pandas: 4, 731, 731 y 17 379. No se ha perdido ninguna fila en la carga.

### 22. Probamos la base de datos

**Qué hacemos:** dos pruebas. Primero intentamos meter un alquiler con una fecha que no existe en `calendario`; después hacemos una primera consulta sobre la vista `v_dia`.

**Por qué la primera prueba:** demuestra que las claves foráneas funcionan. Si MySQL aceptara esa fila, la base admitiría alquileres de días que no existen. Sirve para la demo.

**Por qué `engine.begin()`:** abre una transacción y, si algo falla, la deshace sola. La base queda como estaba.

**La consulta:** por año, cuenta los días, suma los alquileres y calcula la media diaria, el porcentaje de ocasionales y los días de lluvia. `CAST(... AS UNSIGNED)` devuelve las sumas como números enteros: MySQL las da como decimales.

In [ ]:
from sqlalchemy import text
from sqlalchemy.exc import IntegrityError

try:
    with engine.begin() as conexion:   # abre una transacción y la deshace si hay un error
        conexion.execute(text(
            "INSERT INTO alquileres_hora (fecha, hora, situacion_id, ocasionales, registrados, total) "
            "VALUES ('2030-01-01', 8, 1, 1, 1, 2)"
        ))
    print("Cuidado: la clave foránea no se está comprobando.")
except IntegrityError as error:
    print("Rechazado, como debe ser:", error.orig)

Rechazado, como debe ser: (1452, 'Cannot add or update a child row: a foreign key constraint fails (`bikeshare`.`alquileres_hora`, CONSTRAINT `alquileres_hora_ibfk_1` FOREIGN KEY (`fecha`) REFERENCES `calendario` (`fecha`))')


In [ ]:
pd.read_sql_query("""
    SELECT anio,
           COUNT(*)                                       AS dias,
           CAST(SUM(total) AS UNSIGNED)                   AS alquileres,
           ROUND(AVG(total))                              AS media_diaria,
           ROUND(100 * SUM(ocasionales) / SUM(total), 1)  AS pct_ocasionales,
           CAST(SUM(dia_lluvioso) AS UNSIGNED)            AS dias_de_lluvia
    FROM v_dia
    GROUP BY anio
    ORDER BY anio
""", engine)

,anio,dias,alquileres,media_diaria,pct_ocasionales,dias_de_lluvia
0,2011,365,1243103,3406.0,19.9,125
1,2012,366,2049576,5600.0,18.2,116


**Qué leemos:**
- MySQL rechaza el alquiler del 2030-01-01 con el error 1452 (*a foreign key constraint fails*): la clave foránea entre `alquileres_hora` y `calendario` funciona.
- En 2011 hubo 1 243 103 alquileres (3 406 al día de media) y en 2012, 2 049 576 (5 600 al día): un 65 % más en un año.
- Los ocasionales pasan del 19,9 % al 18,2 % del total: los registrados crecen algo más deprisa. Apunta en la dirección de la hipótesis H3; lo miraremos mes a mes con una función de ventana en el notebook de informe.
- Hubo 125 días de lluvia en 2011 y 116 en 2012.

### 23. Guardamos todo en Google Drive

**Qué hacemos:** guardamos en la carpeta del proyecto en Drive:
- `data/raw`: los datos tal como llegaron (Open-Meteo, `day.csv`, `hour.csv`);
- `data/clean`: las cuatro tablas limpias en CSV;
- `sql/01_schema.sql`: el esquema de la base de datos;
- `sql/03_bikeshare_dump.sql`: la base completa (estructura y datos), creada con `mysqldump`.

**Por qué:** la máquina de Colab, con su MySQL, se borra al cerrar la sesión; Drive no. El volcado es lo que pide el enunciado: con él, cualquiera recrea la base en su MySQL o en MySQL Workbench (*Server → Data Import*) sin ejecutar este notebook.

**Por qué `--databases`:** el volcado incluye la instrucción que crea la base `bikeshare`, así se importa de una vez.

**Ojo:** al ejecutar esta celda, Colab puede pedir permiso para acceder a Drive: hay que aceptarlo con la cuenta del proyecto. Si en Drive quedó `data/bikeshare.db` de la versión anterior con SQLite, se puede borrar.

In [ ]:
import shutil
from pathlib import Path
from google.colab import drive

drive.mount("/content/drive")

proyecto = Path("/content/drive/MyDrive/sql-database")
for carpeta in ["data/raw", "data/clean", "sql"]:
    (proyecto / carpeta).mkdir(parents=True, exist_ok=True)

# Datos tal como llegaron
for nombre, tabla in [("weather_daily", weather_df), ("day", bikes_day_df), ("hour", bikes_hour_df)]:
    tabla.to_csv(proyecto / "data/raw" / f"{nombre}.csv", index=False)

# Tablas limpias
for nombre, tabla in tablas:
    tabla.to_csv(proyecto / "data/clean" / f"{nombre}.csv", index=False)

# Esquema y volcado completo de la base (estructura y datos), para abrirlo en MySQL Workbench
shutil.copy("01_schema.sql", proyecto / "sql/01_schema.sql")
ruta_volcado = proyecto / "sql/03_bikeshare_dump.sql"
!mysqldump --databases {DB_NAME} > "{ruta_volcado}"

sorted(str(ruta.relative_to(proyecto)) for ruta in proyecto.rglob("*") if ruta.is_file())

MessageError: Error: credential propagation was unsuccessful

In [ ]:
# Comprobamos el volcado: tamaño, cabecera y cuántas instrucciones INSERT contiene.
print(round(ruta_volcado.stat().st_size / 1024), "KB")
!head -n 5 "{ruta_volcado}"
!grep -c "INSERT INTO" "{ruta_volcado}"

**Qué leemos:**
- En Drive están los tres archivos originales, las cuatro tablas limpias, el esquema y el volcado.
- El volcado ocupa 851 KB y empieza con la cabecera de `mysqldump` (MySQL 8.0, base `bikeshare`): es una exportación de MySQL.
- Tiene 4 instrucciones `INSERT INTO`, una por tabla: `mysqldump` agrupa todas las filas de cada tabla en una sola instrucción larga, que se importa mucho más deprisa que una por fila.

### Conclusión de la parte 2

- La base `bikeshare` en MySQL tiene cuatro tablas y una vista: 4 códigos de tiempo, 731 días de calendario, 731 días de meteorología y 17 379 horas de alquileres.
- Las claves primarias y foráneas funcionan: MySQL rechaza una fecha que no existe.
- 2012 tuvo un 65 % más de alquileres que 2011, y los registrados crecen algo más deprisa que los ocasionales.

**Siguientes pasos:**
1. ERD: importar el volcado en MySQL Workbench y dibujar el diagrama con *Database → Reverse Engineer*.
2. Notebook de informe: consultas SQL (`JOIN`, `GROUP BY`, subconsultas y funciones de ventana) y gráficos.
3. Pasar las funciones de limpieza a un archivo `.py` y los valores fijos (URL, fechas, umbrales, credenciales) a `config.yaml`.